File: integrations/app/src/main/java/app/revanced/bilibili/patches/CommentImagePatch.java
package app.revanced.bilibili.patches;

import android.text.TextUtils;
import android.view.HapticFeedbackConstants;

import androidx.annotation.Keep;

import com.bilibili.lib.imageviewer.fragment.ImageFragment;

import app.revanced.bilibili.settings.Settings;
import app.revanced.bilibili.utils.KtUtils;
import app.revanced.bilibili.utils.Reflex;
import app.revanced.bilibili.utils.Utils;

public class CommentImagePatch {

    private static final int imageViewId = Utils.getResId("image_view", "id");

    @Keep
    public static void bindClickListener(ImageFragment fragment) {
        if (!Settings.SAVE_COMMENT_IMAGE.getBoolean()) return;
        var arguments = fragment.getArguments();
        if (arguments == null || arguments.isEmpty()) return;
        var imageItem = arguments.getParcelable("image_item");
        if (imageItem == null) return;
        var field = Reflex.findFirstFieldByExactTypeOrNull(imageItem.getClass().getSuperclass(), String.class);
        if (field == null) return;
        var imageUrl = (String) Reflex.getFieldValue(imageItem, field);
        if (TextUtils.isEmpty(imageUrl) || !imageUrl.startsWith("http"))
            return;
        var atIndex = imageUrl.indexOf('@');
        if (atIndex != -1)
            imageUrl = imageUrl.substring(0, atIndex);
        var view = fragment.getView();
        if (view == null) return;
        var imageView = view.findViewById(imageViewId);
        if (imageView == null) return;
        final var finalImageUrl = imageUrl;
        imageView.setOnLongClickListener(v -> {
            v.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
            KtUtils.saveImage(finalImageUrl);
            return true;
        });
    }
}

File: integrations/app/src/main/java/app/revanced/bilibili/patches/CommentImagePatch.kt
package app.revanced.bilibili.patches

import android.os.Parcelable
import android.view.HapticFeedbackConstants
import android.view.View
import androidx.annotation.Keep
import app.revanced.bilibili.patches.main.ApplicationDelegate
import app.revanced.bilibili.settings.Settings
import app.revanced.bilibili.utils.*
import com.bilibili.lib.imageviewer.fragment.ImageFragment

object CommentImagePatch {
    @Keep
    @JvmStatic
    fun bindClickListener(fragment: ImageFragment) {
        if (!Settings.SaveCommentImage() && !Versions.ge7_76_0()) return
        val imageItem = fragment.arguments?.parcelable<Parcelable>("image_item") ?: return
        val field = Reflex.findFirstFieldByExactTypeOrNull(
            imageItem.javaClass.superclass, String::class.java
        )?.name ?: return
        val imageUrl = imageItem.getFieldAs<String?>(field).orEmpty().substringBefore('@')
            .takeIf { it.startsWith("http") } ?: return
        fragment.view?.findView<View>("image_view")?.setOnLongClickListener {
            it.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
            saveImage(imageUrl)
            true
        }
    }

    @Keep
    @JvmStatic
    fun onCommentImageClick(pictures: List<*>, index: Int) {
        val imageViewerModelClass = runCatchingOrNull {
            Class.forName("com.bilibili.lib.imageviewer.ImageViewerModel")
        } ?: return
        val picturesField = imageViewerModelClass.declaredFields.first { it.type == List::class.java }
        val indexField = imageViewerModelClass.declaredFields.first { it.type == Int::class.javaPrimitiveType }
        val flagFields = imageViewerModelClass.declaredFields.filter {
            it.type == Boolean::class.javaPrimitiveType
        }.take(2)
        val goMethod = imageViewerModelClass.declaredMethods.first {
            it.returnType == Void.TYPE && it.parameterTypes.isEmpty()
        }
        imageViewerModelClass.new(ApplicationDelegate.getTopActivity()).apply {
            setField(picturesField.name, pictures)
            setField(indexField.name, index)
            flagFields.forEach { setField(it.name, true) }
            callMethod(goMethod.name)
        }
    }
}
